# xi. Model Inference

In [2]:
import pickle
import pandas as pd
import numpy as np
import random

In [3]:
#load model scaler dan model knn yang sudah disimpen dalam bentuk pkl
with open('../models/scaler.pkl', 'rb') as file_1:
  scaler = pickle.load(file_1)

with open('../models/model_knn.pkl', 'rb') as file_2:
  model_knn = pickle.load(file_2)

## Membuat data dummy

In [4]:
# Atur seed di sini
seed_value = 777
random.seed(seed_value)
np.random.seed(seed_value)

# Jumlah data point yang ingin dibuat
n = 100

# Membuat data untuk kolom 'distance'
limit_balance = np.round(np.random.uniform(10000, 800000, n), 2)

# Membuat data untuk kolom 'surge_multiplier'
pay_1 = [-2.0,-1.0,0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0]
pay_1 = random.choices(pay_1, k=n)

pay_2 = [-2.0,-1.0,0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0]
pay_2 = random.choices(pay_2, k=n)

pay_3 = [-2.0,-1.0,0.0,2.0,3.0,4.0,5.0,6.0,7.0]
pay_3 = random.choices(pay_3, k=n)

pay_4 = [-2.0,-1.0,0.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0]
pay_4 = random.choices(pay_4, k=n)

pay_5 = [-2.0,-1.0,0.0,2.0,3.0,4.0,5.0,6.0,7.0]
pay_5 = random.choices(pay_5, k=n)

pay_6 = [-2.0,-1.0,0.0,2.0,3.0,4.0,6.0,7.0]
pay_6 = random.choices(pay_6, k=n)

# Membuat DataFrame
df_inf = pd.DataFrame({
    'limit_balance': limit_balance,
    'pay_1': pay_1,
    'pay_2': pay_2,
    'pay_3': pay_3,
    'pay_4': pay_4,
    'pay_5': pay_5,
    'pay_6': pay_6,
    
})

In [4]:
df_inf

limit_balance  pay_1  pay_2  pay_3  pay_4  pay_5  pay_6
0       130604.35    0.0    7.0    5.0    5.0    2.0    6.0
1       248861.72    2.0   -2.0    3.0    0.0   -2.0    6.0
2        59008.77    2.0   -2.0    3.0    5.0    2.0    7.0
3       373289.67    0.0    7.0    2.0    2.0   -2.0    0.0
4       669850.17    1.0    4.0   -1.0    0.0    7.0    7.0
..            ...    ...    ...    ...    ...    ...    ...
95      772548.37    2.0   -2.0    0.0    0.0    6.0    3.0
96      780417.84    7.0    6.0    7.0    5.0    3.0   -2.0
97      169988.27    4.0    2.0    3.0   -2.0   -1.0    4.0
98      547083.91    4.0    1.0    4.0   -1.0    5.0    2.0
99      377622.60   -2.0    5.0    3.0    7.0    3.0    7.0

[100 rows x 7 columns]

In [5]:
#membagi 2 kolom numerical dan kategorical
df_inf_num = df_inf[['limit_balance']]
df_inf_cat= df_inf[['pay_1', 'pay_2', 'pay_3', 'pay_4','pay_5','pay_6']]

print(df_inf_cat)
print(df_inf_num)

    pay_1  pay_2  pay_3  pay_4  pay_5  pay_6
0     0.0    7.0    5.0    5.0    2.0    6.0
1     2.0   -2.0    3.0    0.0   -2.0    6.0
2     2.0   -2.0    3.0    5.0    2.0    7.0
3     0.0    7.0    2.0    2.0   -2.0    0.0
4     1.0    4.0   -1.0    0.0    7.0    7.0
..    ...    ...    ...    ...    ...    ...
95    2.0   -2.0    0.0    0.0    6.0    3.0
96    7.0    6.0    7.0    5.0    3.0   -2.0
97    4.0    2.0    3.0   -2.0   -1.0    4.0
98    4.0    1.0    4.0   -1.0    5.0    2.0
99   -2.0    5.0    3.0    7.0    3.0    7.0

[100 rows x 6 columns]
    limit_balance
0       130604.35
1       248861.72
2        59008.77
3       373289.67
4       669850.17
..            ...
95      772548.37
96      780417.84
97      169988.27
98      547083.91
99      377622.60

[100 rows x 1 columns]


## Scaling

In [6]:
df_inf_num_scaled = scaler.transform(df_inf_num)
df_inf_num_scaled=pd.DataFrame(df_inf_num_scaled)
df_inf_num_scaled


0
0   0.241209
1   0.477723
2   0.098018
3   0.726579
4   1.319700
..       ...
95  1.525097
96  1.540836
97  0.319977
98  1.074168
99  0.735245

[100 rows x 1 columns]

## Menggabungkan 2 kolom numerikal yang sudah di scaling dan kolom kategorical

In [7]:
df_inf_final = np.concatenate([df_inf_num_scaled,df_inf_cat],axis = 1)
df_inf_final

array([[ 0.2412087 ,  0.        ,  7.        ,  5.        ,  5.        ,
         2.        ,  6.        ],
       [ 0.47772344,  2.        , -2.        ,  3.        ,  0.        ,
        -2.        ,  6.        ],
       [ 0.09801754,  2.        , -2.        ,  3.        ,  5.        ,
         2.        ,  7.        ],
       [ 0.72657934,  0.        ,  7.        ,  2.        ,  2.        ,
        -2.        ,  0.        ],
       [ 1.31970034,  1.        ,  4.        , -1.        ,  0.        ,
         7.        ,  7.        ],
       [ 1.46465534,  6.        ,  0.        ,  7.        ,  6.        ,
         4.        ,  3.        ],
       [ 1.1486426 ,  8.        , -1.        , -2.        ,  5.        ,
         5.        ,  3.        ],
       [ 1.21422404,  8.        ,  6.        ,  5.        ,  4.        ,
         4.        ,  3.        ],
       [ 0.425344  ,  4.        ,  4.        ,  5.        ,  2.        ,
         0.        ,  3.        ],
       [ 1.01756628,  2.    

In [9]:
#membuat kolom predict 
y_pred_inf = model_knn.predict(df_inf_final)
y_pred_inf

array([0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1])

In [10]:
#membuat kolom predict kedalam data frame dan memberi nama kolomnya dengan 'Def_payment - Prediction'
y_pred_inf_df = pd.DataFrame(y_pred_inf, columns=['Def_payment - Prediction'])
y_pred_inf_df

Def_payment - Prediction
0                          0
1                          1
2                          0
3                          1
4                          1
..                       ...
95                         1
96                         1
97                         1
98                         1
99                         1

[100 rows x 1 columns]

In [10]:
#membuat variabel baru dengan menggabungkan 2 tabel yaitu df_inf dan y _preditc inf
tabel_inference = pd.concat([df_inf, y_pred_inf_df], axis=1)
tabel_inference

limit_balance  pay_1  pay_2  pay_3  pay_4  pay_5  pay_6  \
0       130604.35    0.0    7.0    5.0    5.0    2.0    6.0   
1       248861.72    2.0   -2.0    3.0    0.0   -2.0    6.0   
2        59008.77    2.0   -2.0    3.0    5.0    2.0    7.0   
3       373289.67    0.0    7.0    2.0    2.0   -2.0    0.0   
4       669850.17    1.0    4.0   -1.0    0.0    7.0    7.0   
..            ...    ...    ...    ...    ...    ...    ...   
95      772548.37    2.0   -2.0    0.0    0.0    6.0    3.0   
96      780417.84    7.0    6.0    7.0    5.0    3.0   -2.0   
97      169988.27    4.0    2.0    3.0   -2.0   -1.0    4.0   
98      547083.91    4.0    1.0    4.0   -1.0    5.0    2.0   
99      377622.60   -2.0    5.0    3.0    7.0    3.0    7.0   

    Def_payment - Prediction  
0                          0  
1                          1  
2                          0  
3                          1  
4                          1  
..                       ...  
95                         1  
96                         1  
97                         1  
98                         1  
99                         1  

[100 rows x 8 columns]

# xii. Kesimpulan 

- Saya memilih f1 score sebagai parameter score karena menurut saya untuk meminimalisir false postive dan false negative penting .

- Dari hasil std yang mendekati 0 dan berada di bawah mean berarti model KNN yang default nilai STD 0.03 sehingga modelnya best fit . Jadi bisa dikatakan bahwa model knn default memiliki konsistensi yang bagus akan tetapi validasi hasil predictnya kurang bagus dikarenakan hanya sekitar 48%. Dari sisi bisnis model ini belum bisa digunakan karena hasil predictnya yang masih kurang bagus kurang dari 50% sehingga bisa menimbulkan hasil predict yang salah dimana bisa menyebabkan perusahaan rugi.

- Untuk Meningkatkan jumlah user yang bisa bayar ada baiknya untuk penagihan bisa dilakukan ditanggal gajian sekitar 25-31 pada saat gajian karena pada saat tanggal tersebut uang mereka jumlahnya banyak jadi bisa membayar 


## Conceptual Problems :

1. Apakah yang dimaksud dengan coeficient pada logistic regression?

Yang dimaksud dengan coeficient pada logistic regression adalah nilai bobot atau parameter yang digunakan untuk mengukur hubungan antara variabel independen (fitur) dengan variabel dependen (kelas atau target)

2. Apakah fungsi parameter kernel pada SVM? Jelaskan salah satu kernel yang kalian pahami!

Kernel adalah fungsi matematis yang digunakan untuk mengukur kesamaan antara pasangan data dalam ruang fitur
Salah satu kernelnya adalah linear yang digunakan untuk pemisahan data yang dapat dipisahkan dengan garis lurus.

3. Bagaimana cara memilih K yang optimal pada KNN ?

Cara memilih K yang optimal pada KNN adalah dengan membagi data set menjadi 2 yaitu train dan test lalu mendefine model knn. Lalu diolah dengan menggunakan cross validation untuk menentukan parameter mana yang terbaik dan memvariasikan nilai K dan mengamati performa validasi silang, kita bisa menemukan K yang optimal

4. Apa yang dimaksud dengan metrics-metrics berikut : Accuracy, Precision, Recall, F1 Score, dan kapan waktu yang tepat untuk menggunakannya ?


Accuracy: Merupakan rasio dari jumlah prediksi yang benar (positif dan negatif) dibandingkan dengan jumlah total sampel. Berguna ketika distribusi kelas seimbang. Namun, bisa menjadi bias jika kelas tidak seimbang.

Precision: Merupakan rasio dari jumlah prediksi positif yang benar dibandingkan dengan total prediksi positif. Berguna ketika penting untuk menghindari false positive.

Recall: Merupakan rasio dari jumlah prediksi positif yang benar dibandingkan dengan total jumlah sampel positif yang sebenarnya. Berguna ketika penting untuk menghindari false negative.

F1 Score: Merupakan ukuran yang mengkombinasikan precision dan recall. Berguna ketika Anda ingin mencari keseimbangan antara precision dan recall.